core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/FaultTolerantPartitioningSchemeFactory.java

```diff
@@ -54,21 +54,25 @@ public FaultTolerantPartitioningSchemeFactory(NodePartitioningManager nodePartit
 
     public FaultTolerantPartitioningScheme get(PartitioningHandle handle, Optional<Integer> partitionCount)
     {
-        FaultTolerantPartitioningScheme result = cache.get(handle);
+        CacheKey cacheKey = new CacheKey(handle, partitionCount);
+        FaultTolerantPartitioningScheme result = cache.get(cacheKey);
         if (result == null) {
             // Avoid using computeIfAbsent as the "get" method is called recursively from the "create" method
             result = create(handle, partitionCount);
-            cache.put(handle, result);
+            cache.put(cacheKey, result);
         }
-        else if (partitionCount.isPresent()) {
-            // With runtime adaptive partitioning, it's no longer guaranteed that the same handle will always map to
-            // the same partition count. Therefore, use the supplied `partitionCount` as the source of truth.
-            result = result.withPartitionCount(partitionCount.get());
```

Comment:
`withPartitionCount` became unused, remove

Reply:
before the change, we guaranteed that "if value is taken from the cache, the partitionCount is made to be exactly as requested". 
i don't know why we didn't guarantee that for new values. 

maybe we could add this

```java
if (partitionCount.isPresent()) {
  verify(result.getPartitionCount() == partitionCount.get(), "...");
}
```

however, `create` doesn't seem to guarantee this property -- `return new FaultTolerantPartitioningScheme(1, ...`.

Thus this PR (1) fixes caching logic
and introduces a pretty subtle side-effect (maybe correct). If this side effect is intentional, it would be good to call it out (perhaps as separate commit)